bin/propolis-server/src/lib/migrate/source.rs

```diff
@@ -114,77 +117,106 @@ enum RamOfferDiscipline {
-
-    if let Err(err) = proto.run().await {
-        err_tx
-            .send(MigrateSourceCommand::UpdateState(MigrationState::Error))
-            .await
-            .unwrap();
-
-        // We encountered an error, try to inform the remote before bailing
-        // Note, we don't use `?` here as this is a best effort and we don't
-        // want an error encountered during this send to shadow the run error
-        // from the caller.
-        let _ = proto.send_msg(codec::Message::Error(err.clone())).await;
-
-        // If we are capable of setting the dirty bit on guest page table
-        // entries, re-dirty them, so that a later migration attempt can also
-        // only offer dirty pages. If we can't use VM_NPT_OPERATION, a
-        // subsequent migration attempt will offer all pages.
-        //
-        // See the lengthy comment on `RamOfferDiscipline` above for more
-        // details about what's going on here.
-        {
-            let objects = proto.vm.lock_shared().await;
-            for (&GuestAddr(gpa), dirtiness) in proto.dirt.iter().flatten() {
-                if let Err(e) =
-                    objects.vmm_hdl().set_dirty_pages(gpa, dirtiness)
-                {
-                    // Bad news! Our attempt to re-set the dirty bit on these
-                    // pages has failed! Thus, subsequent migration attempts
-                    // /!\ CAN NO LONGER RELY ON DIRTY PAGE TRACKING /!\
-                    // and must always offer all pages in the initial RAM push
-                    // phase.
-                    //
-                    // Record that now so we never try to do this again.
-                    proto
-                        .command_tx
-                        .send(MigrateSourceCommand::RedirtyingFailed)
-                        .await
-                        .unwrap();
-                    // .map_err(|_| MigrateError::StateDriverChannelClosed)?;
+/// The interface to an arbitrary version of the source half of the live
+/// migration protocol.
+//
+// Use `async_trait` here to help generate a `Send` bound on the futures
+// returned by the functions in this trait.
+#[async_trait::async_trait]
+pub(crate) trait SourceProtocol {
+    /// Runs live migration out of the supplied `vm_objects`, writing back any
+    /// state that must be saved for future migration attempts to
+    /// `persistent_state`.
+    ///
+    /// This routine guarantees that the supplied `vm_objects` are paused on
+    /// success and resumed on failure.
+    async fn run(
+        self,
+        vm_objects: &VmObjects,
+        publisher: &mut StatePublisher,
+        persistent_state: &mut PersistentState,
+    ) -> Result<(), MigrateError>;
+}
 
+/// Negotiates a live migration protocol version with a target who has connected
+/// over `conn`. If this is successful, returns a `SourceProtocol`
+/// implementation that can be used to run the requested migration.
+pub(crate) async fn initiate<T: MigrateConn>(
```

Comment:
Most of the logic in this function was taken from the (removed) `source_start` function in `migrate/mod.rs`.